File: app/src/main/java/life/plank/juna/zone/view/adapter/board/match/TimelineAdapter.java
package life.plank.juna.zone.view.adapter.board.match;

import android.content.Context;
import android.support.annotation.NonNull;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;
import life.plank.juna.zone.R;
import life.plank.juna.zone.data.model.LiveTimeStatus;
import life.plank.juna.zone.data.model.MatchEvent;
import life.plank.juna.zone.util.view.BaseRecyclerView;

import static life.plank.juna.zone.util.common.AppConstants.FULL_TIME;
import static life.plank.juna.zone.util.common.AppConstants.GOAL;
import static life.plank.juna.zone.util.common.AppConstants.HALF_TIME;
import static life.plank.juna.zone.util.common.AppConstants.HT;
import static life.plank.juna.zone.util.common.AppConstants.KICK_OFF;
import static life.plank.juna.zone.util.common.AppConstants.LIVE;
import static life.plank.juna.zone.util.common.AppConstants.LIVE_TIME;
import static life.plank.juna.zone.util.common.AppConstants.PENALTY;
import static life.plank.juna.zone.util.common.AppConstants.RED_CARD;
import static life.plank.juna.zone.util.common.AppConstants.SUBSTITUTION;
import static life.plank.juna.zone.util.common.AppConstants.YELLOW_CARD;
import static life.plank.juna.zone.util.common.AppConstants.YELLOW_RED;
import static life.plank.juna.zone.util.common.DataUtil.getFormattedExtraMinutes;
import static life.plank.juna.zone.util.common.DataUtil.isNullOrEmpty;
import static life.plank.juna.zone.util.view.UIDisplayUtil.getDp;

public class TimelineAdapter extends BaseRecyclerView.Adapter<TimelineAdapter.TimelineViewHolder> {

    private List<MatchEvent> matchEventList;
    private Context context;

    public TimelineAdapter(Context context) {
        this.context = context;
        matchEventList = new ArrayList<>();
    }

    @NonNull
    @Override
    public TimelineViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new TimelineViewHolder(
                LayoutInflater.from(parent.getContext()).inflate(R.layout.item_timeline, parent, false),
                this
        );
    }

    @Override
    public int getItemCount() {
        return matchEventList.size();
    }

    public void updateEvents(List<MatchEvent> matchEventList) {
        this.matchEventList.addAll(matchEventList);
        notifyDataSetChanged();
    }

    /**
     * Adding the element at its respective position.
     * Best case runtime : O(1)
     * Worst case runtime : O(n)
     * Live case runtime : O(1) - because live events are going to be coming "live", so it will always insert at 0th position.
     */
    public void updateLiveEvents(List<MatchEvent> matchEventList) {
        if (!isNullOrEmpty(matchEventList) && !isNullOrEmpty(matchEventList.get(0).getPlayerName())) {
            int positionToInsert = 0;
            for (MatchEvent matchEvent : this.matchEventList) {
                if (matchEvent.getMinute() > matchEventList.get(0).getMinute()) {
                    positionToInsert = this.matchEventList.indexOf(matchEvent);
                } else break;
            }
            this.matchEventList.addAll(positionToInsert, matchEventList);
            notifyItemRangeInserted(positionToInsert, matchEventList.size());
        }
    }

    /**
     * Adding the whistle event at its respective position.
     */
    public void updateWhistleEvent(LiveTimeStatus timeStatus) {
        int positionToInsert = 0;
        for (MatchEvent matchEvent : matchEventList) {
            if (matchEvent.getMinute() > timeStatus.getMinute()) {
                positionToInsert = this.matchEventList.indexOf(matchEvent);
            } else break;
        }
        matchEventList.add(positionToInsert, new MatchEvent(timeStatus));
        notifyItemInserted(positionToInsert);
    }

    static class TimelineViewHolder extends BaseRecyclerView.ViewHolder {

        @BindView(R.id.center_space)
        View verticalLine;
        @BindView(R.id.whistle_event_layout)
        RelativeLayout whistleLayout;
        @BindView(R.id.whistle)
        ImageView whistleImage;
        @BindView(R.id.whistle_text_1)
        TextView whistleEventUp;
        @BindView(R.id.whistle_text_2)
        TextView whistleEventDown;
        @BindView(R.id.minute)
        TextView minuteView;
        @BindView(R.id.timeline_layout)
        LinearLayout timelineLayout;
        @BindView(R.id.timeline_text_1)
        TextView timelineEventUp;
        @BindView(R.id.timeline_text_2)
        TextView timelineEventDown;

        private final WeakReference<TimelineAdapter> ref;
        private MatchEvent event;

        TimelineViewHolder(View itemView, TimelineAdapter adapter) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            ref = new WeakReference<>(adapter);
        }

        @Override
        public void bind() {
            event = ref.get().matchEventList.get(getAdapterPosition());

            if (event.getLiveTimeStatus() != null) {
                setLayout(true);
                onWhistleEvent();
            } else {
                setLayout(false);
                placeTimelineLayout();

                switch (event.getEventType()) {
                    case GOAL:
                        onGoalEvent();
                        break;
                    case YELLOW_CARD:
                    case RED_CARD:
                    case YELLOW_RED:
                        onCardEvent();
                        break;
                    case SUBSTITUTION:
                        onSubstitutionEvent();
                        break;
                    case PENALTY:
                        onPenaltyEvent();
                        break;
                    default:
                        break;
                }
            }
        }

        private void setLayout(boolean isWhistleEvent) {
            whistleLayout.setVisibility(isWhistleEvent ? View.VISIBLE : View.GONE);
            timelineLayout.setVisibility(isWhistleEvent ? View.GONE : View.VISIBLE);
            minuteView.setVisibility(isWhistleEvent ? View.GONE : View.VISIBLE);
        }

        private void placeTimelineLayout() {
            RelativeLayout.LayoutParams params = (RelativeLayout.LayoutParams) timelineLayout.getLayoutParams();
            params.addRule(
                    event.isHomeTeam() ? RelativeLayout.ALIGN_PARENT_START : RelativeLayout.ALIGN_PARENT_END,
                    RelativeLayout.TRUE
            );
            params.removeRule(event.isHomeTeam() ? RelativeLayout.ALIGN_PARENT_END : RelativeLayout.ALIGN_PARENT_START);
            params.addRule(event.isHomeTeam() ? RelativeLayout.START_OF : RelativeLayout.END_OF, R.id.center_space);
            params.removeRule(event.isHomeTeam() ? RelativeLayout.END_OF : RelativeLayout.START_OF);
            params.rightMargin = (int) getDp(event.isHomeTeam() ? 22 : 0);
            params.leftMargin = (int) getDp(event.isHomeTeam() ? 0 : 22);
            timelineEventUp.setGravity(Gravity.CENTER_VERTICAL | (event.isHomeTeam() ? Gravity.END : Gravity.START));
            timelineEventDown.setGravity(Gravity.CENTER_VERTICAL | (event.isHomeTeam() ? Gravity.END : Gravity.START));
            LinearLayout.LayoutParams eventParams = (LinearLayout.LayoutParams) timelineEventUp.getLayoutParams();
            eventParams.gravity = Gravity.CENTER_VERTICAL | (event.isHomeTeam() ? Gravity.END : Gravity.START);
            timelineEventUp.setLayoutParams(eventParams);
            timelineEventDown.setLayoutParams(eventParams);
            timelineLayout.setLayoutParams(params);
        }

        private void onWhistleEvent() {
            whistleImage.setVisibility(View.VISIBLE);
            verticalLine.setVisibility(View.INVISIBLE);
            minuteView.setVisibility(View.GONE);
            whistleEventUp.setText(getTimedEventString());
            whistleEventDown.setVisibility(View.VISIBLE);
            String whistleEventString = "+" + getTimedEventExtraString();
            whistleEventDown.setText(whistleEventString);
        }

        private void onCardEvent() {
            setCenterLayout();
            timelineEventDown.setCompoundDrawablePadding((int) getDp(10));
            int suitableCardDrawable = event.getEventType().contains(ref.get().context.getString(R.string.red)) ?
                    event.isHomeTeam() ?
                            R.drawable.red_left :
                            R.drawable.red_right :
                    event.isHomeTeam() ?
                            R.drawable.yellow_left :
                            R.drawable.yellow_right;
            if (event.isHomeTeam()) {
                timelineEventUp.setCompoundDrawablesWithIntrinsicBounds(0, 0, suitableCardDrawable, 0);
            } else {
                timelineEventUp.setCompoundDrawablesWithIntrinsicBounds(suitableCardDrawable, 0, 0, 0);
            }
            timelineEventUp.setText(event.getPlayerName());
            timelineEventDown.setVisibility(View.GONE);
        }

        private void onSubstitutionEvent() {
            setCenterLayout();
            timelineEventDown.setCompoundDrawablePadding((int) getDp(10));
            if (event.isHomeTeam()) {
                timelineEventUp.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_substitute_in, 0);
                timelineEventDown.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_substitute_out, 0);
            } else {
                timelineEventUp.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_substitute_in, 0, 0, 0);
                timelineEventDown.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_substitute_out, 0, 0, 0);
            }
            timelineEventUp.setText(event.getPlayerName());
            timelineEventDown.setVisibility(View.VISIBLE);
            timelineEventDown.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12);
            timelineEventDown.setText(event.getRelatedPlayerName());
        }

        private void onGoalEvent() {
            setCenterLayout();
            timelineEventDown.setCompoundDrawablePadding((int) getDp(4));
            timelineEventUp.setCompoundDrawablesWithIntrinsicBounds(
                    event.isHomeTeam() ? 0 : R.drawable.ic_goal_right,
                    0,
                    event.isHomeTeam() ? R.drawable.ic_goal_left : 0,
                    0
            );
            timelineEventUp.setText(event.getPlayerName());
            timelineEventDown.setVisibility(View.VISIBLE);
            timelineEventDown.setTextSize(TypedValue.COMPLEX_UNIT_SP, 10);
            String timelineEventDownText;
            if (event.getRelatedPlayerName() != null) {
                timelineEventDown.setCompoundDrawablesWithIntrinsicBounds(
                        event.isHomeTeam() ? R.drawable.ic_assist : 0,
                        0,
                        event.isHomeTeam() ? 0 : R.drawable.ic_assist,
                        0
                );
                timelineEventDownText = event.isHomeTeam() ?
                        event.getRelatedPlayerName() + "    " + event.getResult() :
                        event.getResult() + "    " + event.getRelatedPlayerName();
            } else {
                timelineEventDown.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
                timelineEventDownText = event.getResult();
            }
            timelineEventDown.setText(timelineEventDownText);
        }

        private void onPenaltyEvent() {
            setCenterLayout();
            timelineEventDown.setVisibility(View.VISIBLE);
            timelineEventUp.setVisibility(View.VISIBLE);
            timelineEventDown.setTextSize(TypedValue.COMPLEX_UNIT_SP, 10);
            timelineEventUp.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
            timelineEventDown.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0);
            timelineEventUp.setText(event.getPlayerName());
            timelineEventDown.setText(PENALTY);
        }

        private String getTimedEventString() {
            switch (Objects.requireNonNull(event.getLiveTimeStatus()).getTimeStatus()) {
                case LIVE:
                    return KICK_OFF;
                case HT:
                    return HALF_TIME;
                default:
                    return FULL_TIME;
            }
        }

        private String getTimedEventExtraString() {
            switch (Objects.requireNonNull(event.getLiveTimeStatus()).getTimeStatus()) {
                case LIVE:
                    return LIVE_TIME;
                default:
                    return getFormattedExtraMinutes(event.getLiveTimeStatus().getExtraMinute());
            }
        }

        private void setCenterLayout() {
            whistleImage.setVisibility(View.GONE);
            verticalLine.setVisibility(View.VISIBLE);
            minuteView.setVisibility(View.VISIBLE);
            setMinute();
        }

        private void setMinute() {
            String minuteText = event.getMinute() + "'";
            minuteView.setText(minuteText);
        }
    }
}

File: app/src/main/java/life/plank/juna/zone/ui/board/adapter/match/TimelineAdapter.kt
package life.plank.juna.zone.ui.board.adapter.match

import android.util.TypedValue
import android.view.*
import android.widget.*
import androidx.recyclerview.widget.RecyclerView
import com.prembros.facilis.util.isNullOrEmpty
import kotlinx.android.synthetic.main.item_timeline.view.*
import life.plank.juna.zone.R
import life.plank.juna.zone.data.model.football.*
import life.plank.juna.zone.service.CommonDataService.findString
import life.plank.juna.zone.service.MatchDataService
import life.plank.juna.zone.util.common.AppConstants.*
import life.plank.juna.zone.util.view.UIDisplayUtil.getDp
import java.util.*

class TimelineAdapter : RecyclerView.Adapter<TimelineAdapter.TimelineViewHolder>() {

    private val matchEventList: MutableList<MatchEvent>

    init {
        matchEventList = ArrayList()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TimelineViewHolder =
            TimelineViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_timeline, parent, false))

    override fun onBindViewHolder(holder: TimelineViewHolder, position: Int) {
        holder.run {
            val event = matchEventList[position]

            if (event.liveTimeStatus != null) {
                setLayout(true)
                onWhistleEvent()
            } else {
                setLayout(false)
                placeTimelineLayout()

                when (event.eventType) {
                    GOAL -> onGoalEvent(false)
                    YELLOW_CARD, RED_CARD, YELLOW_RED -> onCardEvent()
                    SUBSTITUTION -> onSubstitutionEvent()
                    PENALTY -> onPenaltyEvent(false)
                    MISSED_PENALTY -> onPenaltyEvent(true)
                    OWN_GOAL -> onGoalEvent(true)
                    else -> {
                    }
                }
            }
        }
    }

    override fun getItemCount(): Int = matchEventList.size

    fun updateEvents(matchEventList: List<MatchEvent>) {
        this.matchEventList.addAll(matchEventList)
        notifyDataSetChanged()
    }

    private fun TimelineViewHolder.setLayout(isWhistleEvent: Boolean) {
        itemView.run {
            whistle_event_layout.visibility = if (isWhistleEvent) View.VISIBLE else View.GONE
            timeline_layout.visibility = if (isWhistleEvent) View.GONE else View.VISIBLE
            minute.visibility = if (isWhistleEvent) View.GONE else View.VISIBLE
        }
    }

    private fun TimelineViewHolder.placeTimelineLayout() {
        itemView.run {
            val params = timeline_layout.layoutParams as RelativeLayout.LayoutParams
            params.addRule(if (event.isHomeTeam) RelativeLayout.ALIGN_PARENT_START else RelativeLayout.ALIGN_PARENT_END, RelativeLayout.TRUE)
            params.removeRule(if (event.isHomeTeam) RelativeLayout.ALIGN_PARENT_END else RelativeLayout.ALIGN_PARENT_START)
            params.addRule(if (event.isHomeTeam) RelativeLayout.START_OF else RelativeLayout.END_OF, R.id.center_space)
            params.removeRule(if (event.isHomeTeam) RelativeLayout.END_OF else RelativeLayout.START_OF)
            params.rightMargin = getDp((if (event.isHomeTeam) 22 else 0).toFloat()).toInt()
            params.leftMargin = getDp((if (event.isHomeTeam) 0 else 22).toFloat()).toInt()
            timeline_event_up.gravity = Gravity.CENTER_VERTICAL or if (event.isHomeTeam) Gravity.END else Gravity.START
            timeline_event_down.gravity = Gravity.CENTER_VERTICAL or if (event.isHomeTeam) Gravity.END else Gravity.START
            val eventParams = timeline_event_up.layoutParams as LinearLayout.LayoutParams
            eventParams.gravity = Gravity.CENTER_VERTICAL or if (event.isHomeTeam) Gravity.END else Gravity.START
            timeline_event_up.layoutParams = eventParams
            timeline_event_down.layoutParams = eventParams
            timeline_layout.layoutParams = params
        }
    }

    private fun TimelineViewHolder.onWhistleEvent() {
        itemView.run {
            whistle.visibility = View.VISIBLE
            center_space.visibility = View.INVISIBLE
            minute.visibility = View.GONE
            whistle_event_up.text = getTimedEventString()
            whistle_event_down.visibility = View.VISIBLE
            whistle_event_down.text = getTimedEventExtraString()
        }
    }

    private fun TimelineViewHolder.onCardEvent() {
        itemView.run {
            setCenterLayout()
            timeline_event_down.compoundDrawablePadding = getDp(10f).toInt()
            val suitableCardDrawable = if (event.eventType.contains(findString(R.string.red)))
                if (event.isHomeTeam)
                    R.drawable.red_left
                else
                    R.drawable.red_right
            else if (event.isHomeTeam)
                R.drawable.yellow_left
            else
                R.drawable.yellow_right
            if (event.isHomeTeam) {
                timeline_event_up.setCompoundDrawablesWithIntrinsicBounds(0, 0, suitableCardDrawable, 0)
            } else {
                timeline_event_up.setCompoundDrawablesWithIntrinsicBounds(suitableCardDrawable, 0, 0, 0)
            }
            timeline_event_up.text = event.playerName
            timeline_event_down.visibility = View.GONE
        }
    }

    private fun TimelineViewHolder.onSubstitutionEvent() {
        itemView.run {
            setCenterLayout()
            timeline_event_down.compoundDrawablePadding = getDp(10f).toInt()
            if (event.isHomeTeam) {
                timeline_event_up.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_substitute_in, 0)
                timeline_event_down.setCompoundDrawablesWithIntrinsicBounds(0, 0, R.drawable.ic_substitute_out, 0)
            } else {
                timeline_event_up.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_substitute_in, 0, 0, 0)
                timeline_event_down.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_substitute_out, 0, 0, 0)
            }
            timeline_event_up.text = event.playerName
            timeline_event_down.visibility = View.VISIBLE
            timeline_event_down.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12f)
            timeline_event_down.text = event.relatedPlayerName
        }
    }

    private fun TimelineViewHolder.onGoalEvent(isOwnGoal: Boolean) {
        itemView.run {
            setCenterLayout()
            timeline_event_down.compoundDrawablePadding = getDp(4f).toInt()
            timeline_event_up.setCompoundDrawablesWithIntrinsicBounds(
                    if (event.isHomeTeam) 0 else R.drawable.ic_goal_right,
                    0,
                    if (event.isHomeTeam) R.drawable.ic_goal_left else 0,
                    0
            )
            timeline_event_up.text = event.playerName
            timeline_event_down.visibility = View.VISIBLE
            timeline_event_down.setTextSize(TypedValue.COMPLEX_UNIT_SP, 10f)
            val timelineEventDownText = StringBuilder()
            if (isOwnGoal && event.isHomeTeam) {
                timelineEventDownText.append(OWN_GOAL_LOWERCASE).append(COMMA)
            }
            if (event.relatedPlayerName != null) {
                timeline_event_down.setCompoundDrawablesWithIntrinsicBounds(
                        if (event.isHomeTeam) R.drawable.ic_assist else 0,
                        0,
                        if (event.isHomeTeam) 0 else R.drawable.ic_assist,
                        0
                )
                timelineEventDownText.append(if (event.isHomeTeam) event.relatedPlayerName else event.result)
                        .append(WIDE_SPACE)
                        .append(if (event.isHomeTeam) event.result else event.relatedPlayerName)
            } else {
                timeline_event_down.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
                timelineEventDownText.append(event.result)
            }
            if (isOwnGoal && !event.isHomeTeam) {
                timelineEventDownText.append(COMMA).append(OWN_GOAL_LOWERCASE)
            }
            timeline_event_down.text = timelineEventDownText
        }
    }

    private fun TimelineViewHolder.onPenaltyEvent(isMissed: Boolean) {
        itemView.run {
            setCenterLayout()
            timeline_event_down.visibility = View.VISIBLE
            timeline_event_up.visibility = View.VISIBLE
            timeline_event_down.setTextSize(TypedValue.COMPLEX_UNIT_SP, 10f)
            timeline_event_up.setCompoundDrawablesWithIntrinsicBounds(
                    if (isMissed) 0 else if (event.isHomeTeam) 0 else R.drawable.ic_goal_right,
                    0,
                    if (isMissed) 0 else if (event.isHomeTeam) R.drawable.ic_goal_left else 0,
                    0
            )
            timeline_event_down.setCompoundDrawablesWithIntrinsicBounds(0, 0, 0, 0)
            timeline_event_up.text = event.playerName
            val penaltyText = StringBuilder()
            if (isMissed) {
                penaltyText.append(findString(R.string.missed_penalty))
            } else {
                penaltyText.append(if (event.isHomeTeam) findString(R.string.penalty) else event.result)
                        .append(COMMA)
                        .append(if (event.isHomeTeam) event.result else findString(R.string.penalty))
            }
            timeline_event_down.text = penaltyText
        }
    }

    private fun TimelineViewHolder.setCenterLayout() {
        itemView.run {
            whistle.visibility = View.GONE
            center_space.visibility = View.VISIBLE
            minute.visibility = View.VISIBLE
            setMinute()
        }
    }

    private fun TimelineViewHolder.setMinute() {
        itemView.run {
            val minuteText = event.minute.toString() + "'"
            minute.text = minuteText
        }
    }

    private fun TimelineViewHolder.getTimedEventString(): String {
        itemView.run {
            return when (Objects.requireNonNull<LiveTimeStatus>(event.liveTimeStatus).timeStatus) {
                LIVE -> KICK_OFF
                HT -> HALF_TIME
                else -> FULL_TIME
            }
        }
    }

    private fun TimelineViewHolder.getTimedEventExtraString(): String {
        itemView.run {
            return when (Objects.requireNonNull<LiveTimeStatus>(event.liveTimeStatus).timeStatus) {
                LIVE -> LIVE_TIME
                else -> MatchDataService.getFormattedExtraMinutes(event.liveTimeStatus!!.extraMinute)
            }
        }
    }

    /**
     * Adding the element at its respective position.
     * Best case runtime : O(1)
     * Worst case runtime : O(n)
     * Live case runtime : O(1) - because live events are going to be coming "live", so it will always insert at 0th position.
     */
    fun updateLiveEvents(matchEventList: List<MatchEvent>) {
        if (!isNullOrEmpty(matchEventList) && !isNullOrEmpty(matchEventList[0].playerName)) {
            var positionToInsert = 0
            for (matchEvent in this.matchEventList) {
                if (matchEvent.minute > matchEventList[0].minute) {
                    positionToInsert = this.matchEventList.indexOf(matchEvent)
                } else
                    break
            }
            this.matchEventList.addAll(positionToInsert, matchEventList)
            notifyItemRangeInserted(positionToInsert, matchEventList.size)
        }
    }

    /**
     * Adding the whistle event at its respective position.
     */
    fun updateWhistleEvent(timeStatus: LiveTimeStatus) {
        var positionToInsert = 0
        for (matchEvent in matchEventList) {
            if (matchEvent.minute > timeStatus.minute) {
                positionToInsert = this.matchEventList.indexOf(matchEvent)
            } else
                break
        }
        matchEventList.add(positionToInsert, MatchEvent(timeStatus))
        notifyItemInserted(positionToInsert)
    }

    class TimelineViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        lateinit var event: MatchEvent
    }
}